Source/WebCore/editing/FrameSelection.cpp

```diff
@@ -2229,16 +2249,15 @@ void FrameSelection::updateAppearance()
     // If the caret moved, stop the blink timer so we can restart with a
     // black caret in the new location.
     if (caretRectChangedOrCleared || !shouldBlink || shouldStopBlinkingDueToTypingCommand(m_document.get()))
-        m_caretBlinkTimer.stop();
+        caretAnimator().stop();
 
     // Start blinking with a black caret. Be sure not to restart if we're
     // already blinking in the right location.
-    if (shouldBlink && !m_caretBlinkTimer.isActive()) {
-        if (Seconds blinkInterval = RenderTheme::singleton().caretBlinkInterval())
-            m_caretBlinkTimer.startRepeating(blinkInterval);
+    if (shouldBlink && !caretAnimator().isActive()) {
+        caretAnimator().start(m_document->domWindow()->nowTimestamp());
 
-        if (!m_caretPaint) {
-            m_caretPaint = true;
+        if (!caretAnimator().isVisible()) {
+            caretAnimator().setVisible(true);
             invalidateCaretRect();
```

Comment:
Shouldn't it be the responsibility of the `caretAnimator` to call `caretAnimationDidUpdate` if visibility changes, which then triggers the invalidation? Ideally there'd only be one place that calls `invalidateCaretRect()`.